httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/AsyncHttpRequestRetryExec.java

```diff
@@ -0,0 +1,187 @@
+            public void handleInformationResponse(final HttpResponse response) throws HttpException, IOException {
+                asyncExecCallback.handleInformationResponse(response);
+            }
+
+            @Override
+            public void completed() {
+                if (state.retrying) {
+                    state.execCount++;
+                    try {
+                        internalExecute(state, request, entityProducer, scope, chain, asyncExecCallback);
+                    } catch (final IOException | HttpException ex) {
+                        asyncExecCallback.failed(ex);
+                    }
+                } else {
+                    asyncExecCallback.completed();
+                }
+            }
+
+            @Override
+            public void failed(final Exception cause) {
+                if (cause instanceof IOException) {
+                    final HttpRoute route = scope.route;
+                    final HttpClientContext clientContext = scope.clientContext;
+                    if (entityProducer != null && !entityProducer.isRepeatable()) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("{}: cannot retry non-repeatable request", exchangeId);
+                        }
+                    } else if (retryStrategy.retryRequest(request, (IOException) cause, state.execCount, clientContext)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("{}: {}", exchangeId, cause.getMessage(), cause);
+                        }
+                        if (log.isInfoEnabled()) {
+                            log.info("Recoverable I/O exception ({}) caught when processing request to {}",
+                                    cause.getClass().getName(), route);
+                        }
+                        scope.execRuntime.discardEndpoint();
+                        if (entityProducer != null) {
+                            entityProducer.releaseResources();
+                        }
+                        state.retrying = true;
+                        state.execCount++;
+                        try {
+                            internalExecute(state, request, entityProducer, scope, chain, asyncExecCallback);
+                        } catch (final IOException | HttpException ex) {
+                            asyncExecCallback.failed(ex);
+                        }
+                        return;
+                    }
+                }
+                asyncExecCallback.failed(cause);
+            }
+
+        });
+
+    }
+
+    @Override
+    public void execute(
+            final HttpRequest request,
+            final AsyncEntityProducer entityProducer,
+            final AsyncExecChain.Scope scope,
+            final AsyncExecChain chain,
+            final AsyncExecCallback asyncExecCallback) throws HttpException, IOException {
+        final State state = new State();
+        state.execCount = 1;
```

Comment:
Not very OO, use setters or a ctor?

Reply:
That's a private class, consider it like a record or value class. I see no benefit.

Reply:
Hm, ok.